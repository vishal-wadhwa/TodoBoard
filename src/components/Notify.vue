<template>
  <v-snackbar v-show='!!msg' v-model='snackbar' bottom auto-height>
    <v-icon :color='type' class='mr-3' medium>{{TYPE_TO_ICON[type]}}</v-icon>
    {{msg}}
    <v-btn color='pink' flat @click='snackbar = false'>Close</v-btn>
  </v-snackbar>
</template>
<script>
const TYPES = ['success', 'info', 'warning', 'error']

export default {
  props: {
    msg: String,
    type: {
      type: String,
      default: 'info',
      validator: v => TYPES.includes(v)
    }
  },
  data () {
    return {
      snackbar: false,
      TYPE_TO_ICON: {
        success: 'check_circle',
        info: 'info',
        warning: 'warning',
        error: 'error'
      }
    }
  },
  created () {
    this.snackbar = !!this.msg
  },
  watch: {
    msg (nv) {
      this.snackbar = true
    }
  }
}
</script>
