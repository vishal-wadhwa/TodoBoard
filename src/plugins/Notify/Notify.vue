<template>
  <transition name='v-snack-transition'>
    <div v-show='isActive' class='v-snack' :class='classes' :on='$listeners'>
      <div class='v-snack__wrapper' :style='setBackgroundColor(color, {}).style'>
        <div class='v-snack__content'>
          <v-icon :color='type' class='mr-3' medium>{{TYPE_TO_ICON[type]}}</v-icon>
          {{msg}}
          <v-btn color='pink' flat @click='onClose'>Close</v-btn>
        </div>
      </div>
    </div>
  </transition>
</template>
<script>
import { VSnackbar } from 'vuetify/lib'
import { EventBus } from './event_bus'

export default {
  name: 'Notify',
  extends: VSnackbar,
  data () {
    return {
      type: 'info',
      msg: '',
      TYPE_TO_ICON: {
        success: 'check_circle',
        info: 'info',
        warning: 'warning',
        error: 'error'
      }
    }
  },
  methods: {
    notify (data = {}) {
      if (!data.msg) return
      this.onClose()
      this.msg = data.msg
      this.type = data.type || 'info'
      // watch on this.isActive will trigger this.setTimeout()
      this.$nextTick(() => (this.isActive = true))
    },
    onClose () {
      this.isActive = false
      window.clearTimeout(this.activeTimeout)
    },
    setTimeout () {
      if (!this.isActive) return
      this.activeTimeout = window.setTimeout(this.onClose, this.timeout)
    }
  },
  created () {
    EventBus.$on('show', this.notify)
  }
}
</script>
