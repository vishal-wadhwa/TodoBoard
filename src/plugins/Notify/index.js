import Notify from './Notify'
import { EventBus } from './event_bus'

const Notif = {
  install (Vue, args = {}) {
    if (this.installed) {
      return
    }

    this.installed = true
    this.params = args

    Vue.component(args.componentName || 'notify', Notify)

    const notify = (params) => {
      if (typeof params === 'object') {
        EventBus.$emit('show', params)
      }
    }

    const name = args.name || 'notify'

    Vue.prototype['$' + name] = notify
    Vue[name] = notify
  }
}

export default Notif
