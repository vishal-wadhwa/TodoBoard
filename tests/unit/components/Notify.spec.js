import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import Notify from '@/components/Notify'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

describe('AppLogo.vue', () => {
  // required to vuetify [warn] in snapshot test
  // const app = document.createElement('div')
  // app.setAttribute('data-app', true)
  // document.body.appendChild(app)

  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(Notify, opts) : shallowMount(Notify, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  const msg = 'This is a test notification.'
  const TYPES = ['info', 'success', 'warning', 'error']

  it('does not show notification by default', () => {
    const wrapper = factory({
      localVue
    })

    expect(wrapper.isVisible()).toBeFalsy()
  })

  it('does not show notification if msg prop is empty', () => {
    const wrapper = factory({
      localVue,
      propsData: { msg: '' }
    })

    expect(wrapper.isVisible()).toBeFalsy()
  })

  it('shows notification with appropriate msg', () => {
    const wrapper = factory({
      propsData: { msg },
      localVue
    })

    expect(wrapper.text()).toMatch(msg)
  })

  it('uses info icon by default', () => {
    const wrapper = factory({
      propsData: { msg },
      localVue
    })

    setTimeout(() => expect(wrapper.find({ name: 'v-icon' }).text()).toMatch('info'), 0)
  })

  // Can't get these two cases to run :/
  // it('shows the notification for 6000ms', () => {
  //   jest.setTimeout(6500)
  //   const wrapper = factory({
  //     localVue,
  //     propsData: { msg },
  //     sync: false
  //   })

  //   expect(wrapper.vm.$data.snackbar).toBeTruthy()
  //   setTimeout(() => {
  //     expect(wrapper.text()).toMatch(msg)
  //     console.log('hhahd')
  //     setTimeout(() => {
  //       console.log('hhhelo')
  //       expect(wrapper.vm.$data.snackbar).toBeFalsy()
  //     }, 6100)
  //   }, 0)
  // })

  // it('hides notification when Close button is clicked', done => {
  //   const wrapper = factory({
  //     localVue,
  //     propsData: { msg },
  //     sync: false
  //   })

  //   expect(wrapper.vm.$data.snackbar).toBeTruthy()
  //   wrapper.vm.$nextTick(() => {
  //     wrapper.find({ name: 'v-btn' }).trigger('click')
  //     setTimeout(() => {
  //       expect(wrapper.vm.$data.snackbar).toBeFalsy()
  //       done()
  //     }, 0)
  //   })
  // })

  it('validates type prop to be one of the accepted values', () => {
    const wrapper = factory({
      propsData: { msg },
      localVue
    })

    const typeProp = wrapper.vm.$options.props.type

    expect(typeProp.validator('dalkdfh')).toBeFalsy()
    for (const type of TYPES) {
      expect(typeProp.validator(type)).toBeTruthy()
    }
  })

  it(`should match snapshot with no msg`, () => {
    const wrapper = factory({
      localVue
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  for (const type of TYPES) {
    it(`should match snapshot with type prop as ${type}`, () => {
      const wrapper = factory({
        localVue,
        propsData: { msg, type }
      }, true)

      expect(wrapper.html()).toMatchSnapshot()
    })
  }
})
