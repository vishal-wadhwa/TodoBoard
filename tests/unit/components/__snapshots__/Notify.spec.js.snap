// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLogo.vue should match snapshot with no msg 1`] = `undefined`;

exports[`AppLogo.vue should match snapshot with type prop as error 1`] = `
<div class="v-snack v-snack--active v-snack--auto-height v-snack--bottom" name="v-snack-transition">
  <div class="v-snack__wrapper">
    <div class="v-snack__content"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light error--text" style="font-size: 28px;">error</i>
      This is a test notification.
      <button type="button" class="v-btn v-btn--flat theme--light pink--text">
        <div class="v-btn__content">Close</div>
      </button></div>
  </div>
</div>
`;

exports[`AppLogo.vue should match snapshot with type prop as info 1`] = `
<div class="v-snack v-snack--active v-snack--auto-height v-snack--bottom" name="v-snack-transition">
  <div class="v-snack__wrapper">
    <div class="v-snack__content"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light info--text" style="font-size: 28px;">info</i>
      This is a test notification.
      <button type="button" class="v-btn v-btn--flat theme--light pink--text">
        <div class="v-btn__content">Close</div>
      </button></div>
  </div>
</div>
`;

exports[`AppLogo.vue should match snapshot with type prop as success 1`] = `
<div class="v-snack v-snack--active v-snack--auto-height v-snack--bottom" name="v-snack-transition">
  <div class="v-snack__wrapper">
    <div class="v-snack__content"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light success--text" style="font-size: 28px;">check_circle</i>
      This is a test notification.
      <button type="button" class="v-btn v-btn--flat theme--light pink--text">
        <div class="v-btn__content">Close</div>
      </button></div>
  </div>
</div>
`;

exports[`AppLogo.vue should match snapshot with type prop as warning 1`] = `
<div class="v-snack v-snack--active v-snack--auto-height v-snack--bottom" name="v-snack-transition">
  <div class="v-snack__wrapper">
    <div class="v-snack__content"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light warning--text" style="font-size: 28px;">warning</i>
      This is a test notification.
      <button type="button" class="v-btn v-btn--flat theme--light pink--text">
        <div class="v-btn__content">Close</div>
      </button></div>
  </div>
</div>
`;
